sound/soc/sof/ipc.c

```diff
@@ -672,9 +672,49 @@ int snd_sof_ipc_set_get_comp_data(struct snd_sof_control *scontrol,
 	struct sof_ipc_fw_ready *ready = &sdev->fw_ready;
 	struct sof_ipc_fw_version *v = &ready->version;
 	struct sof_ipc_ctrl_data_params sparams;
+	struct snd_sof_widget *swidget;
+	bool widget_found = false;
 	size_t send_bytes;
 	int err;
 
+	list_for_each_entry_reverse(swidget, &sdev->widget_list, list) {
+		if (swidget->comp_id == scontrol->comp_id) {
+			widget_found = true;
+			break;
+		}
+	}
+
+	if (!widget_found) {
+		dev_err(sdev->dev, "error: can't find widget with id %d\n", scontrol->comp_id);
+		return -EINVAL;
+	}
+
+	/*
+	 * return 0 for SET commands as the cached value will be sent to the DSP when the widget
+	 * has been set up. Return error for GET to notify the userspace that the widget
+	 * data is inaccessible.
+	 */
+	mutex_lock(&swidget->use_count_mutex);
```

Comment:
Upon more thought, all other logic around this is protected by the pcm_mutex, so this kcontrol usage is the only one not covered. 

My votes go to removing the use_count_mutex and having a plain check for use_count here. In worst case, if a some other process is busylooping on the kcontrol and PCM gets open/closed, user-space will get an error (as the IPC fails). This would seem the simplest option.

Reply:
@kv2019i ack. I've removed the mutex and added the constraint that volatile controls can only belong to static pipeline widgets